Openssl v4.0.0 and RC_2_0 build errors

I had some build errors on one file. Here they are

```
[134/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/web_peer_connection.cpp.o
[135/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/dht_storage.cpp.o
[136/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/pe_crypto.cpp.o
[137/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/torrent.cpp.o
FAILED: [code=1] CMakeFiles/torrent-rasterbar.dir/src/torrent.cpp.o
/usr/local/bin/x86_64-linux-musl-g++ -DBOOST_ASIO_ENABLE_CANCELIO -DBOOST_ASIO_HAS_STD_CHRONO -DBOOST_ASIO_NO_DEPRECATED -DBOOST_EXCEPTION_DISABLE -DOPENSSL_NO_DTLS1 -DOPENSSL_NO_SSL2 -DOPENSSL_NO_SSL3 -DOPENSSL_NO_TLS1 -DOPENSSL_NO_TLS1_1 -DTORRENT_BUILDING_LIBRARY -DTORRENT_NO_DEPRECATE -DTORRENT_SSL_PEERS -DTORRENT_USE_LIBCRYPTO -DTORRENT_USE_OPENSSL -D_FILE_OFFSET_BITS=64 -I/root/qbt-build/libtorrent/include -I/root/qbt-build/libtorrent/deps/try_signal -isystem /root/qbt-build/boost -I/usr/include/fortify -I/root/qbt-build/include -Og -g -pipe -fdata-sections -ffunction-sections -fPIC -flto=auto -fno-fat-lto-objects -fstack-clash-protection -fstack-protector-strong -fno-plt -fno-delete-null-pointer-checks -fno-strict-overflow -fno-strict-aliasing -ftrivial-auto-var-init=zero -fexceptions -fstrict-flex-arrays=3 -w -std=c++20 -pthread -static --static -O3 -DNDEBUG -std=gnu++20 -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Wextra -Wpedantic -Wvla -Wno-noexcept-type -Wno-format-zero-length -ftemplate-depth=512 -fexceptions -MD -MT CMakeFiles/torrent-rasterbar.dir/src/torrent.cpp.o -MF CMakeFiles/torrent-rasterbar.dir/src/torrent.cpp.o.d -o CMakeFiles/torrent-rasterbar.dir/src/torrent.cpp.o -c /root/qbt-build/libtorrent/src/torrent.cpp
/root/qbt-build/libtorrent/src/torrent.cpp: In member function 'bool libtorrent::torrent::verify_peer_cert(bool, boost::asio::ssl::verify_context&)':
/root/qbt-build/libtorrent/src/torrent.cpp:1633:35: error: invalid use of incomplete type 'ASN1_IA5STRING' {aka 'struct asn1_string_st'}
 1633 |                         if (domain->type != V_ASN1_IA5STRING || !domain->data || !domain->length) continue;
      |                                   ^~
In file included from /root/qbt-build/include/openssl/crypto.h:38,
                 from /root/qbt-build/include/openssl/comp.h:25,
                 from /root/qbt-build/include/openssl/ssl.h:31,
                 from /root/qbt-build/libtorrent/include/libtorrent/ssl.hpp:64,
                 from /root/qbt-build/libtorrent/include/libtorrent/tracker_manager.hpp:69,
                 from /root/qbt-build/libtorrent/include/libtorrent/alert_types.hpp:69,
                 from /root/qbt-build/libtorrent/include/libtorrent/piece_picker.hpp:60,
                 from /root/qbt-build/libtorrent/include/libtorrent/peer_list.hpp:48,
                 from /root/qbt-build/libtorrent/include/libtorrent/torrent.hpp:67,
                 from /root/qbt-build/libtorrent/src/torrent.cpp:68:
/root/qbt-build/include/openssl/types.h:57:16: note: forward declaration of 'ASN1_IA5STRING' {aka 'struct asn1_string_st'}
   57 | typedef struct asn1_string_st ASN1_INTEGER;
      |                ^~~~~~~~~~~~~~
/root/qbt-build/libtorrent/src/torrent.cpp:1633:72: error: invalid use of incomplete type 'ASN1_IA5STRING' {aka 'struct asn1_string_st'}
 1633 |                         if (domain->type != V_ASN1_IA5STRING || !domain->data || !domain->length) continue;
      |                                                                        ^~
/root/qbt-build/include/openssl/types.h:57:16: note: forward declaration of 'ASN1_IA5STRING' {aka 'struct asn1_string_st'}
   57 | typedef struct asn1_string_st ASN1_INTEGER;
      |                ^~~~~~~~~~~~~~
/root/qbt-build/libtorrent/src/torrent.cpp:1633:89: error: invalid use of incomplete type 'ASN1_IA5STRING' {aka 'struct asn1_string_st'}
 1633 |                         if (domain->type != V_ASN1_IA5STRING || !domain->data || !domain->length) continue;
      |                                                                                         ^~
/root/qbt-build/include/openssl/types.h:57:16: note: forward declaration of 'ASN1_IA5STRING' {aka 'struct asn1_string_st'}
   57 | typedef struct asn1_string_st ASN1_INTEGER;
      |                ^~~~~~~~~~~~~~
/root/qbt-build/libtorrent/src/torrent.cpp:1634:88: error: invalid use of incomplete type 'ASN1_IA5STRING' {aka 'struct asn1_string_st'}
 1634 |                         auto const* torrent_name = reinterpret_cast<char const*>(domain->data);
      |                                                                                        ^~
/root/qbt-build/include/openssl/types.h:57:16: note: forward declaration of 'ASN1_IA5STRING' {aka 'struct asn1_string_st'}
   57 | typedef struct asn1_string_st ASN1_INTEGER;
      |                ^~~~~~~~~~~~~~
/root/qbt-build/libtorrent/src/torrent.cpp:1635:87: error: invalid use of incomplete type 'ASN1_IA5STRING' {aka 'struct asn1_string_st'}
 1635 |                         auto const name_length = aux::numeric_cast<std::size_t>(domain->length);
      |                                                                                       ^~
/root/qbt-build/include/openssl/types.h:57:16: note: forward declaration of 'ASN1_IA5STRING' {aka 'struct asn1_string_st'}
   57 | typedef struct asn1_string_st ASN1_INTEGER;
      |                ^~~~~~~~~~~~~~
/root/qbt-build/libtorrent/src/torrent.cpp:1657:56: error: invalid conversion from 'const X509_NAME*' {aka 'const X509_name_st*'} to 'X509_NAME*' {aka 'X509_name_st*'} [-fpermissive]
 1657 |                 X509_NAME* name = X509_get_subject_name(cert);
      |                                   ~~~~~~~~~~~~~~~~~~~~~^~~~~~
      |                                                        |
      |                                                        const X509_NAME* {aka const X509_name_st*}
/root/qbt-build/libtorrent/src/torrent.cpp:1662:74: error: invalid conversion from 'const X509_NAME_ENTRY*' {aka 'const X509_name_entry_st*'} to 'X509_NAME_ENTRY*' {aka 'X509_name_entry_st*'} [-fpermissive]
 1662 |                         X509_NAME_ENTRY* name_entry = X509_NAME_get_entry(name, i);
      |                                                       ~~~~~~~~~~~~~~~~~~~^~~~~~~~~
      |                                                                          |
      |                                                                          const X509_NAME_ENTRY* {aka const X509_name_entry_st*}
/root/qbt-build/libtorrent/src/torrent.cpp:1663:63: error: invalid conversion from 'const ASN1_STRING*' {aka 'const asn1_string_st*'} to 'ASN1_STRING*' {aka 'asn1_string_st*'} [-fpermissive]
 1663 |                         common_name = X509_NAME_ENTRY_get_data(name_entry);
      |                                       ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
      |                                                               |
      |                                                               const ASN1_STRING* {aka const asn1_string_st*}
/root/qbt-build/libtorrent/src/torrent.cpp:1665:47: error: invalid use of incomplete type 'ASN1_STRING' {aka 'struct asn1_string_st'}
 1665 |                 if (common_name && common_name->data && common_name->length)
      |                                               ^~
/root/qbt-build/include/openssl/types.h:57:16: note: forward declaration of 'ASN1_STRING' {aka 'struct asn1_string_st'}
   57 | typedef struct asn1_string_st ASN1_INTEGER;
      |                ^~~~~~~~~~~~~~
/root/qbt-build/libtorrent/src/torrent.cpp:1665:68: error: invalid use of incomplete type 'ASN1_STRING' {aka 'struct asn1_string_st'}
 1665 |                 if (common_name && common_name->data && common_name->length)
      |                                                                    ^~
/root/qbt-build/include/openssl/types.h:57:16: note: forward declaration of 'ASN1_STRING' {aka 'struct asn1_string_st'}
   57 | typedef struct asn1_string_st ASN1_INTEGER;
      |                ^~~~~~~~~~~~~~
/root/qbt-build/libtorrent/src/torrent.cpp:1667:93: error: invalid use of incomplete type 'ASN1_STRING' {aka 'struct asn1_string_st'}
 1667 |                         auto const* torrent_name = reinterpret_cast<char const*>(common_name->data);
      |                                                                                             ^~
/root/qbt-build/include/openssl/types.h:57:16: note: forward declaration of 'ASN1_STRING' {aka 'struct asn1_string_st'}
   57 | typedef struct asn1_string_st ASN1_INTEGER;
      |                ^~~~~~~~~~~~~~
/root/qbt-build/libtorrent/src/torrent.cpp:1668:92: error: invalid use of incomplete type 'ASN1_STRING' {aka 'struct asn1_string_st'}
 1668 |                         auto const name_length = aux::numeric_cast<std::size_t>(common_name->length);
      |                                                                                            ^~
/root/qbt-build/include/openssl/types.h:57:16: note: forward declaration of 'ASN1_STRING' {aka 'struct asn1_string_st'}
   57 | typedef struct asn1_string_st ASN1_INTEGER;
      |                ^~~~~~~~~~~~~~
[138/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/msg.cpp.o
[139/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/item.cpp.o
[140/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/smart_ban.cpp.o
[141/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/dos_blocker.cpp.o
[142/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/ut_pex.cpp.o
[143/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/ut_metadata.cpp.o
[144/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/ed25519/add_scalar.cpp.o
[145/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/node_entry.cpp.o
[146/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/node_id.cpp.o
[147/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/get_item.cpp.o
[148/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/find_data.cpp.o
[149/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/dht_tracker.cpp.o
[150/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/get_peers.cpp.o
[151/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/put_data.cpp.o
[152/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/rpc_manager.cpp.o
[153/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/sample_infohashes.cpp.o
[154/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/refresh.cpp.o
[155/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/node.cpp.o
[156/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/traversal_algorithm.cpp.o
[157/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/kademlia/routing_table.cpp.o
[158/168] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/session_impl.cpp.o
ninja: build stopped: subcommand failed.
```

I gave them to gemini pro and I got this patch and the errors were resolved.

**_**AI generated patch:**_**

```diff
diff --git a/src/torrent.cpp b/src/torrent.cpp
index c6564b6..7610f67 100644
--- a/src/torrent.cpp
+++ b/src/torrent.cpp
@@ -1630,9 +1630,15 @@ aux::vector<download_priority_t, piece_index_t> file_to_piece_prio(
 			GENERAL_NAME* gen = sk_GENERAL_NAME_value(gens, i);
 			if (gen->type != GEN_DNS) continue;
 			ASN1_IA5STRING* domain = gen->d.dNSName;
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+			if (ASN1_STRING_type(domain) != V_ASN1_IA5STRING || !ASN1_STRING_get0_data(domain) || !ASN1_STRING_length(domain)) continue;
+			auto const* torrent_name = reinterpret_cast<char const*>(ASN1_STRING_get0_data(domain));
+			auto const name_length = aux::numeric_cast<std::size_t>(ASN1_STRING_length(domain));
+#else
 			if (domain->type != V_ASN1_IA5STRING || !domain->data || !domain->length) continue;
 			auto const* torrent_name = reinterpret_cast<char const*>(domain->data);
 			auto const name_length = aux::numeric_cast<std::size_t>(domain->length);
+#endif
 
 #ifndef TORRENT_DISABLE_LOGGING
 			if (i > 1) names += " | n: ";
@@ -1654,18 +1660,29 @@ aux::vector<download_priority_t, piece_index_t> file_to_piece_prio(
 
 		// no match in the alternate names, so try the common names. We should only
 		// use the "most specific" common name, which is the last one in the list.
-		X509_NAME* name = X509_get_subject_name(cert);
+		auto name = X509_get_subject_name(cert);
 		int i = -1;
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+		ASN1_STRING const* common_name = nullptr;
+#else
 		ASN1_STRING* common_name = nullptr;
+#endif
 		while ((i = X509_NAME_get_index_by_NID(name, NID_commonName, i)) >= 0)
 		{
-			X509_NAME_ENTRY* name_entry = X509_NAME_get_entry(name, i);
+			auto name_entry = X509_NAME_get_entry(name, i);
 			common_name = X509_NAME_ENTRY_get_data(name_entry);
 		}
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+		if (common_name && ASN1_STRING_get0_data(common_name) && ASN1_STRING_length(common_name))
+		{
+			auto const* torrent_name = reinterpret_cast<char const*>(ASN1_STRING_get0_data(common_name));
+			auto const name_length = aux::numeric_cast<std::size_t>(ASN1_STRING_length(common_name));
+#else
 		if (common_name && common_name->data && common_name->length)
 		{
 			auto const* torrent_name = reinterpret_cast<char const*>(common_name->data);
 			auto const name_length = aux::numeric_cast<std::size_t>(common_name->length);
+#endif
 
 #ifndef TORRENT_DISABLE_LOGGING
 			if (!names.empty()) names += " | n: ";
```

It's not too big a deal right now since qtbase does not support it yet either but this fix seems simple enough to get v4 support?

--- a/src/torrent.cpp
+++ b/src/torrent.cpp
@@ -1630,9 +1630,9 @@ aux::vector<download_priority_t, piece_index_t> file_to_piece_prio(
 			GENERAL_NAME* gen = sk_GENERAL_NAME_value(gens, i);
 			if (gen->type != GEN_DNS) continue;
 			ASN1_IA5STRING* domain = gen->d.dNSName;
-			if (domain->type != V_ASN1_IA5STRING || !domain->data || !domain->length) continue;
-			auto const* torrent_name = reinterpret_cast<char const*>(domain->data);
-			auto const name_length = aux::numeric_cast<std::size_t>(domain->length);
+			if (ASN1_STRING_type(domain) != V_ASN1_IA5STRING || !ASN1_STRING_get0_data(domain) || !ASN1_STRING_length(domain)) continue;
+			auto const* torrent_name = reinterpret_cast<char const*>(ASN1_STRING_get0_data(domain));
+			auto const name_length = aux::numeric_cast<std::size_t>(ASN1_STRING_length(domain));
 
 #ifndef TORRENT_DISABLE_LOGGING
 			if (i > 1) names += " | n: ";
@@ -1654,18 +1654,22 @@ aux::vector<download_priority_t, piece_index_t> file_to_piece_prio(
 
 		// no match in the alternate names, so try the common names. We should only
 		// use the "most specific" common name, which is the last one in the list.
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+		X509_NAME const* name = X509_get_subject_name(cert);
+#else
 		X509_NAME* name = X509_get_subject_name(cert);
+#endif
 		int i = -1;
-		ASN1_STRING* common_name = nullptr;
+		ASN1_STRING const* common_name = nullptr;
 		while ((i = X509_NAME_get_index_by_NID(name, NID_commonName, i)) >= 0)
 		{
-			X509_NAME_ENTRY* name_entry = X509_NAME_get_entry(name, i);
+			X509_NAME_ENTRY const* name_entry = X509_NAME_get_entry(name, i);
 			common_name = X509_NAME_ENTRY_get_data(name_entry);
 		}
-		if (common_name && common_name->data && common_name->length)
+		if (common_name && ASN1_STRING_get0_data(common_name) && ASN1_STRING_length(common_name))
 		{
-			auto const* torrent_name = reinterpret_cast<char const*>(common_name->data);
-			auto const name_length = aux::numeric_cast<std::size_t>(common_name->length);
+			auto const* torrent_name = reinterpret_cast<char const*>(ASN1_STRING_get0_data(common_name));
+			auto const name_length = aux::numeric_cast<std::size_t>(ASN1_STRING_length(common_name));
 
 #ifndef TORRENT_DISABLE_LOGGING
 			if (!names.empty()) names += " | n: ";

